File: sdk/flutter/example/contract_smoke.dart
// Contract smoke for the Flutter/Dart SDK.
//
// Run from the repo root against a seeded backend (see backend/scripts/seed_sdk_contract.py):
//
//     cd sdk/flutter && EXPERIMENTLY_API_KEY=... dart run example/contract_smoke.dart
//
// Needs the Flutter toolchain (the package depends on the `flutter` SDK); run
// `flutter pub get` once first so `dart run` does not print resolution output.
//
// Environment: EXPERIMENTLY_API_URL (default http://localhost:8000), EXPERIMENTLY_API_KEY
// (required), CONTRACT_EXPERIMENT_KEY (default sdk_contract_ab), CONTRACT_FLAG_KEY (default
// sdk_contract_flag), CONTRACT_USER_ID (default smoke-<uuid>).
//
// Prints exactly one JSON line on stdout and exits 0; on failure prints one line to stderr and
// exits 1. Uses only the SDK's public (Flutter-free) API.

import 'dart:convert';
import 'dart:io';
import 'dart:math';

import 'package:experimentation_sdk/experimentation_sdk_core.dart';

String _env(String name, String defaultValue) {
  final value = Platform.environment[name];
  return (value == null || value.isEmpty) ? defaultValue : value;
}

Never _fail(String message) {
  stderr.writeln('flutter contract smoke: ${message.replaceAll('\n', ' ')}');
  exit(1);
}

/// Random RFC 4122 version-4 UUID (no `uuid` package needed).
String _uuidV4() {
  final rng = Random.secure();
  final bytes = List<int>.generate(16, (_) => rng.nextInt(256));
  bytes[6] = (bytes[6] & 0x0f) | 0x40; // version 4
  bytes[8] = (bytes[8] & 0x3f) | 0x80; // RFC 4122 variant
  final hex = bytes.map((b) => b.toRadixString(16).padLeft(2, '0')).join();
  return '${hex.substring(0, 8)}-${hex.substring(8, 12)}-${hex.substring(12, 16)}-'
      '${hex.substring(16, 20)}-${hex.substring(20)}';
}

Future<void> main() async {
  try {
    await _run();
  } catch (e) {
    _fail('$e');
  }
  exit(0);
}

Future<void> _run() async {
  final apiUrl = _env('EXPERIMENTLY_API_URL', 'http://localhost:8000');
  final apiKey = _env('EXPERIMENTLY_API_KEY', '');
  if (apiKey.isEmpty) _fail('EXPERIMENTLY_API_KEY is required');
  final experimentKey = _env('CONTRACT_EXPERIMENT_KEY', 'sdk_contract_ab');
  final flagKey = _env('CONTRACT_FLAG_KEY', 'sdk_contract_flag');
  final userId = _env('CONTRACT_USER_ID', 'smoke-${_uuidV4()}');

  final config = SdkConfig(
    apiKey: apiKey,
    baseUrl: apiUrl,
    timeout: const Duration(seconds: 10),
    offlineFallback: false,
  );
  final attributes = <String, dynamic>{'source': 'contract_smoke', 'sdk': 'flutter'};

  // `client` keeps the assignment and flag cached (needed for the fan-out step);
  // `fresh` starts with an empty cache so the second assignment is a real round-trip
  // that proves the assignment is sticky on the server.
  final client = ExperimentationClient(config: config);
  final fresh = ExperimentationClient(config: config);
  await client.init();
  await fresh.init();

  try {
    // 1. Assign twice → identical (sticky on the server).
    final first = await client.getAssignment(experimentKey, userId, attributes: attributes);
    if (first == null) {
      _fail('assignment failed for $experimentKey (not ACTIVE, bad key, or API unreachable)');
    }
    final second = await fresh.getAssignment(experimentKey, userId, attributes: attributes);
    if (second == null) _fail('second assignment failed for $experimentKey');
    if (!const {'control', 'treatment'}.contains(first.variantName)) {
      _fail("unexpected variant_name '${first.variantName}' for $experimentKey");
    }
    final sticky = first.variantName == second.variantName && first.variantId == second.variantId;
    if (!sticky) {
      _fail('assignment not sticky: ${first.variantName}/${first.variantId} '
          'vs ${second.variantName}/${second.variantId}');
    }

    // 2. Evaluate the flag → enabled is a bool (seeded flag is 100% on).
    //    evaluateFlag never throws, so a failure shows up as "nothing cached".
    final flag = await client.evaluateFlag(flagKey, userId);
    if (!client.getEvaluatedFlags(userId).contains(flagKey)) {
      _fail('flag evaluation failed for $flagKey (not ACTIVE, bad key, or API unreachable)');
    }

    // 3. Track `purchase` with a value and the experiment key → POST /tracking/track.
    final trackOk = await client.track(
      'purchase',
      userId,
      experimentKey: experimentKey,
      value: 12.5,
      properties: {'sdk': 'flutter', 'currency': 'USD'},
    );
    if (!trackOk) _fail('track purchase failed');

    // 4. Track `page_view` without a key → POST /tracking/batch fanned out to the
    //    cached assignment + flag.
    if (client.getAssignments(userId).length != 1) {
      _fail('expected one cached assignment before fan-out');
    }
    final fanoutOk = await client.track('page_view', userId, properties: {'page': '/smoke'});
    if (!fanoutOk) _fail('fan-out track failed');

    // The SDK exposes a batch call, so also send a 2-event batch.
    final batchOk = await client.trackBatch([
      TrackEvent(userId: userId, eventName: 'add_to_cart', experimentKey: experimentKey, value: 1),
      TrackEvent(userId: userId, eventName: 'checkout', featureFlagKey: flagKey),
    ]);
    if (!batchOk) _fail('batch track failed');

    stdout.writeln(jsonEncode({
      'sdk': 'flutter',
      'assign': {
        'variant_name': first.variantName,
        'is_control': first.isControl,
        'sticky': sticky,
      },
      'flag': {'enabled': flag.enabled},
      'track': {'ok': trackOk},
      'fanout': {'ok': fanoutOk && batchOk},
    }));
  } finally {
    await client.close();
    await fresh.close();
  }
}

File: sdk/edge/examples/contract_smoke.mjs
/**
 * Contract smoke for the Edge SDK (see docs/sdk-guide.md, "Contract smoke").
 *
 * Runs under plain Node >= 18 — the SDK only uses Web-standard globals
 * (`fetch`, `AbortController`, `Headers`), which Node provides, so no edge
 * runtime adapter is needed.
 *
 * Run from the repo root:
 *   cd sdk/edge && npm run build --silent && EXPERIMENTLY_API_KEY=... node examples/contract_smoke.mjs
 *
 * Env: EXPERIMENTLY_API_URL (default http://localhost:8000), EXPERIMENTLY_API_KEY (required),
 *      CONTRACT_EXPERIMENT_KEY (default sdk_contract_ab), CONTRACT_FLAG_KEY (default sdk_contract_flag),
 *      CONTRACT_USER_ID (default smoke-<uuid>).
 *
 * Prints exactly one JSON line on stdout and exits 0; on failure prints one line to stderr and exits 1.
 */
import { randomUUID } from 'node:crypto';
import { EdgeExperimentationClient } from '../dist/index.js';

const apiUrl = process.env.EXPERIMENTLY_API_URL || 'http://localhost:8000';
const apiKey = process.env.EXPERIMENTLY_API_KEY;
const experimentKey = process.env.CONTRACT_EXPERIMENT_KEY || 'sdk_contract_ab';
const flagKey = process.env.CONTRACT_FLAG_KEY || 'sdk_contract_flag';
const userId = process.env.CONTRACT_USER_ID || `smoke-${randomUUID()}`;

function fail(message) {
  process.stderr.write(`edge contract smoke failed: ${message}\n`);
  process.exit(1);
}

if (!apiKey) fail('EXPERIMENTLY_API_KEY is required');

// Track requests with a wrapped fetch so we can tell "swallowed failure" from "ok":
// the SDK's track() never throws, so failures only show up on the wire.
const trackFailures = [];
const fetchImpl = async (url, init) => {
  const response = await fetch(url, init);
  if (typeof url === 'string' && url.includes('/api/v1/tracking/') && !url.endsWith('/assign') && !response.ok) {
    trackFailures.push(`${url} -> ${response.status}`);
  }
  return response;
};

// A generous timeout: the edge default (500 ms) is tuned for production latency.
const client = new EdgeExperimentationClient({ apiKey, baseUrl: apiUrl, timeout: 5000, fetch: fetchImpl });

try {
  // 1. Assign twice. clearCache() between the calls forces the second one to hit the
  //    server, so "sticky" reflects the server's answer rather than the local cache.
  const first = await client.getAssignment(experimentKey, userId, { source: 'contract-smoke' });
  if (!first) fail('assignment failed (null)');
  client.clearCache();
  const second = await client.getAssignment(experimentKey, userId);
  if (!second) fail('second assignment failed (null)');

  if (!['control', 'treatment'].includes(first.variantName)) {
    fail(`unexpected variant_name ${JSON.stringify(first.variantName)}`);
  }
  const sticky = first.variantName === second.variantName && first.variantId === second.variantId;
  if (!sticky) fail(`assignment not sticky: ${first.variantName} then ${second.variantName}`);

  // 2. Evaluate the flag (100% on in the seeded data).
  const flag = await client.evaluateFlag(flagKey, userId);
  if (typeof flag.enabled !== 'boolean') fail('flag.enabled is not a boolean');
  if (client.evaluateFlagSync(flagKey, userId) !== flag.enabled) fail('sync cache read disagrees with evaluateFlag');

  // 3. Track with an explicit experiment key.
  await client.track('purchase', userId, { source: 'contract-smoke' }, { value: 12.5, experimentKey });
  if (trackFailures.length) fail(`track: ${trackFailures[0]}`);

  // 4. Track without a key: fans out to the cached assignment + evaluated flag.
  if (client.getAssignments(userId).length !== 1 || client.getEvaluatedFlags(userId).length !== 1) {
    fail('cache does not hold one assignment and one flag; fan-out would send nothing');
  }
  await client.track('page_view', userId, { page: '/contract-smoke' });
  if (trackFailures.length) fail(`fan-out: ${trackFailures[0]}`);

  // 4b. Explicit two-event batch.
  const batch = await client.trackBatch([
    { eventName: 'add_to_cart', userId, experimentKey, value: 1 },
    { eventName: 'flag_seen', userId, featureFlagKey: flagKey },
  ]);
  if (batch.failureCount > 0 || batch.successCount !== 2) fail(`batch: ${JSON.stringify(batch)}`);

  process.stdout.write(
    JSON.stringify({
      sdk: 'edge',
      assign: { variant_name: second.variantName, is_control: second.isControl, sticky },
      flag: { enabled: flag.enabled },
      track: { ok: true },
      fanout: { ok: true },
    }) + '\n',
  );
} catch (err) {
  fail(err && err.message ? err.message : String(err));
}
